README.md
# Read this before getting the program for the first time (or when updating the program)

Current version: v2.10

Required Python version: v3.6+

Check the bottom of this README for required packages* (only required for screens below 1080p)

To get this program follow this:

https://help.github.com/en/github/creating-cloning-and-archiving-repositories/cloning-a-repository

Or just click the green button called "Code" and download the project's zip.

#### Please take note if you have used this before
- Duplicate your results.txt file before using versions v2.xx to be sure that marks do not get corrupted.

---

#### If you are downloading this project for the first time follow the steps below:
1. Rename results_example.json to results.json (or duplicate results_example.json and then rename)
1. Open Command Prompt or Terminal
3. Navigate to the directory where you have downloaded the project file
4. Run the following code next: "python3 results.py"
5. Enter your marks into the respective courses (in decimal or quotient format -- i.e. 0.96 or 24/25 -- 
if you only have percentages, divide the value by 100)
   
You will have to re-run the results.py program if you want to see the updated marks in the "Current Marks for the semester" screen.

#### If you are coming from v1.0x follow the steps below:
1. Open Command Prompt or Terminal
2. Navigate to the directory where you have downloaded the project file
3. Run the following code: "python3 EditMarks.py" (you should only run this program once)
4. Then run the following code: "python3 results.py"
5. Use the program like normal

You will have to re-run the results.py program if you want to see the updated marks in the "Current Marks for the semester" screen.

#### If you are coming from v0.0x follow the steps below:
1. Open Command Prompt or Terminal
2. Navigate to the directory where you have downloaded the project file
3. Run the following code: "python3 make_results_file.py" (you must only run this program once)
4. Then run the following code: "python3 results.py"
5. Use the program like normal

You will have to re-run the results.py program if you want to see the updated marks in the "Current Marks for the semester" screen.

#### *Users on computers with screen resolution below 1080p

1. Open Command Prompt or Terminal
2. Run "pip3 install pyautogui"
3. Use the program like normal

## Keeping your program up-to-date: How-to
Download the latest version of the results.py and replace the current file in your directory with the latest 
versions.

(Make sure that results.json is in the same directory as the new results.py otherwise the program will not run as expected)

Then just run the program.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](https://keepachangelog.com/)
and this project adheres to [Semantic Versioning](https://semver.org/).

## [v2.10] - Minor Update - 2021-09-20

### Added
- Semester GPA calculation
- Ability to generate a progress report listing GPA and marks for each course, as well as the overview of the degree so far

### Changed
- Nothing

### Fixed
- Nothing

## [v2.09] - Minor Update - 2021-09-19

### Added
- CON2026S course mark calculation and assessments

### Changed
- Nothing

### Fixed
- Nothing

## [v2.08] - Minor Update - 2021-08-15

### Added
- Term and degree GPA calculations

### Changed
- Nothing

### Fixed
- Nothing

## [v2.07] - Minor Update - 2021-08-02

### Added
- Mark forms for 2nd year 2nd semester courses
- Formulas for EEE2044S, EEE2047S, MAM2084S, and PHY2010S

### Changed
- Adjusted course layout alphabetically
- Added a space between WPS and the week number in PHY courses

### Fixed
- Nothing

## [v2.06] - Minor Update - 2021-07-19

### Added
- Nothing

### Changed
- Added "Maximum Marks" to the "Current Marks for Semester" page

### Fixed
- Nothing

## [v2.05] - Minor Update - 2021-06-30

### Added
- Nothing

### Changed
- Removed Tutorial Test 7 for MEC1009F
- Added Tutorial Total for MAM2083F 
  (in case the user wants to use the mark calculated by the lecturer)
- MAM2083F Formula
- MEC1009F Formula

### Fixed
- Nothing

## [v2.04] - Minor Update - 2021-06-21

### Added
- Practical and pre-practical 3 for EEE2045F

### Changed
- Removed Quiz 11 through 13 for MAM2083F

### Fixed
- MAM2083F Formula
- EEE2045F Formula

## [v2.03] - Minor Update - 2021-05-31

### Added
- Nothing

### Changed
- Nothing

### Fixed
- MAM2083F Formula


## [v2.02] - Minor Update - 2021-05-19

### Added
- Correct formula for EEE2045F

### Changed
- Nothing

### Fixed
- EEE2045F was not being marked

## [v2.01] - Minor Update - 2021-05-17

### Added
- Maximum marks possible has been added to each course

### Changed
- Nothing

### Fixed
- Nothing

## [v2.00] - Minor Update - 2021-03-16

### Added
- Formula for EEE2045F, EEE2046F, and MEC1009F
- All known assessments to their respective courses*

### Changed
- Nothing

### Fixed
- Nothing

*How the practical percentage of the course mark for EEE2045F is calculated is unclear - TBD

## [v2.00c] - Minor Update - 2021-03-13

### Added
- Formula for EEE2048F
- 4 EEE2048F practical tasks, application task, academic writing task, MCQ, 
  report, capstone proposal, and capstone report addition JSON file

### Changed
- Nothing

### Fixed
- Nothing

## [v2.00b] - Minor Update - 2021-03-11

### Added
- Formula for MAM2083F
- 13 MAM2083F quizzes, 2 class tests, and 1 exam

### Changed
- JSON now automatically updates itself with new assignments for courses 

### Fixed
- Nothing

## [v2.00a] - Major Update - 2021-03-08

### Added
- Course view template for easy addition of courses
- 2nd year courses w/o formulas (awaiting Notes to Students)

### Changed
- Using JSON for storing marks
- Made the mark calculation function cleaner
- Re-coded all the first year course classes to use the course view template
- StartPage now lists each year of the degree which has either been or is being completed
- Adjusted the README to reflect the updated version

### Fixed
- Many minor bugs

## [v1.09] - Minor Update - 2020-11-23

### Added
- Nothing

### Changed
- Nothing

### Fixed
- Added the final test for MAM1021S back into the formula 

## [v1.08] - Minor Update - 2020-11-22

### Added
- Nothing

### Changed
- Adjusted the MAM1021S formula to use average of top 8 quizzes

### Fixed
- Nothing 

## [v1.07] - Minor Update - 2020-11-16

### Added
- GUI window sizing support for screen with resolutions below 1080p

### Changed
- README.md needed some grammar adjustments

### Fixed
- Results labels now overwrite old ones instead of being plastered on top of the old label

## [v1.06] - Minor Update - 2020-11-09

### Added
- Quiz 10 for MAM1021S

### Changed
- Added a little clarification to README.md

### Fixed
- Quiz 10 for MAM1021S was missing from results.txt
 
## [v1.05] - Minor Update - 2020-11-04

### Added
- Changelog
- Two Milestones

### Changed
- Made all calls to results.txt refer to a variable containing 'results.txt'.

### Fixed
- MAM1021S had 0-5% remaining after all marks had been entered 
  depending on the whether the mark was 0% or 100% on the final test.


make_results_file.py
import os


def adjustForSecondSemester(f):
    with open(f, mode='r') as file:
        lines = file.readlines()

        if 'ADJUSTED\n' not in lines:
            lines.insert(0, 'ADJUSTED\n')
            os.remove('results.txt')
            new_file = open('results.txt', 'w')

            courses = ['mam1020f', 'phy1012f', 'eee1006f', 'csc1015f', 'mec1003f']

            for line in lines:
                if any(substring in line for substring in courses):
                    lines[lines.index(line)] = ''
                elif 'mam' in line:
                    lines[lines.index(line)] = line.replace('mam', 'mam1020f')
                elif 'phy' in line:
                    lines[lines.index(line)] = line.replace('phy', 'phy1012f')
                elif 'eee' in line:
                    lines[lines.index(line)] = line.replace('eee', 'eee1006f')
                elif 'csc' in line:
                    lines[lines.index(line)] = line.replace('csc', 'csc1015f')
                elif 'mec' in line:
                    lines[lines.index(line)] = line.replace('mec', 'mec1003f')

            lines = [line for line in lines if line]

            for i in range(1, 10):
                lines.append(f'mam1021s_quiz_{i}:TBA:0\n')

            lines += ['mam1021s_test_1:TBA:0\n', 'mam1021s_test_2:TBA:0\n', 'mam1021s_test_3:TBA:0\n',
                      'mam1021s_final_test:TBA:0\n']

            for i in range(1, 13):
                lines.append(f'phy1013s_wps_{i}:TBA:0\n')

            lines += ['phy1013s_lab_1:TBA:0\n', 'phy1013s_lab_2:TBA:0\n', 'phy1013s_lab_3:TBA:0\n',
                      'phy1013s_lab_4:TBA:0\n', 'phy1013s_lab_test:TBA:0\n', 'phy1013s_test_1:TBA:0\n',
                      'phy1013s_test_2:TBA:0\n', 'phy1013s_test_3:TBA:0\n', 'phy1013s_test_4:TBA:0\n']

            for i in range(1, 13):
                lines.append(f'csc1016s_quiz_{i}:TBA:0\n')

            lines += ['csc1016s_practical_test_1:TBA:0\n', 'csc1016s_practical_test_2:TBA:0\n',
                      'csc1016s_theory_test_1:TBA:0\n', 'csc1016s_theory_test_2:TBA:0\n',
                      'csc1016s_theory_test_3:TBA:0\n']

            for i in range(1, 7):
                lines.append(f'csc1016s_assignment_{i}:TBA:0\n')

            lines += ['eee1007s_test_1:TBA:0\n', 'eee1007s_test_2:TBA:0\n', 'eee1007s_assignment:TBA:0\n']

            for i in range(1, 8):
                lines.append(f'axl1200s_reflection_piece_{i}:TBA:0\n')

            lines += ['axl1200s_research_project:TBA:0\n']

            new_file.writelines(lines)
        else:
            print("You've already run this program")


if __name__ == '__main__':
    adjustForSecondSemester('results.txt')


EditMarks.py
#  Copyright (c) 2020 David Young.
#  All rights reserved
#


def csc1016s_quizzes(file):
    lines = []
    with open(file, "r") as openfile:
        temp_lines = openfile.readlines()

        for line in temp_lines:
            if "csc1016s_quiz_" in line and eval(line[line.rfind("_") + 1:line.find(":")]) > 7:
                pass
            else:
                lines.append(line)

    with open(file, "w") as writefile:
        writefile.writelines(lines)


def mam1021s_quizzes(file):
    lines = []
    with open(file, "r") as openfile:
        temp_lines = openfile.readlines()

        for line in temp_lines:
            if "mam1021s_quiz_" in line and eval(line[line.rfind("_") + 1:line.find(":")]) == 9:
                lines.append(line)
                lines.append("mam1021s_quiz_10:TBA:0\n") if "mam1021s_quiz_10" not in temp_lines[temp_lines.index(line) + 1] else None
            else:
                lines.append(line)

    with open(file, "w") as writefile:
        writefile.writelines(lines)


if __name__ == '__main__':
    # 21 Oct 2020
    csc1016s_quizzes(file="results.txt")
    # 09 Nov 2020
    mam1021s_quizzes(file="results.txt")
